enhancements/installer/agent-based-installation-in-hive.md

```diff
@@ -0,0 +1,647 @@
+service can generate and embed some token which it can later validate.
+
+#### Use of metal3
+
+[metal3](https://metal3.io/) will be used to boot the assisted discovery ISO
+for bare-metal deployments. Specifically, the BareMetalHost API will be used
+from a Hub cluster to boot the discovery ISO on hosts that will be used to
+form new clusters.
+
+For this first iteration, it is assumed that the hub cluster itself is using
+the bare metal platform, and thus will have baremetal-operator installed
+and available for use. In the future it may be desirable to make the metal3
+capabilities available on hub clusters that are not using the bare metal
+platform.
+
+A separate [enhancement to
+metal3](https://github.com/metal3-io/metal3-docs/pull/150) proposes a new
+capability in the BareMetalHost API enabling it to boot live ISOs other than
+the Ironic Python Agent. That feature is required so that automation in a
+cluster can boot the discovery ISO on known hardware as the first step toward
+provisioning that hardware.
+
+#### Host Approval
+
+It is important that when an agent checks in, it not be allowed to join a
+cluster until an actor has approved it. From a security standpoint, it is not
+OK for anyone who can access a copy of a discovery ISO and/or access the API
+where agents report themselves to implicitly have the capability to join a
+cluster as a Node.
+
+The Agent CRD will have a field in which to designate that it is approved. If
+the host was booted via the baremetal-operator, approval will be granted
+automatically. The same automation that caused the host to boot would have the
+ability to recognize the resulting Agent and mark it as approved.
+
+#### Namespaces and Multi-tenancy
+
+While we expect a ClusterDeployment to have its own namespace (which will be
+increasingly valuable once we have central machine management and need to run
+different machine-api components per cluster), Agents will be created in a
+separate namespace where their InstallEnv exists. "Late binding" in particular
+is a scenario where Agents will exist prior to a cluster. And of course a
+collection of Agents may be created in advance of allocating them to various
+clusters.
+
+Thus when a Cluster Creator creates a namespace and a new ClusterDeployment,
+that ClusterDeployment will need to consume Agents from a different namespace.
+But the Infra Owner may not want them to have the ability to consume any Agent
+from any namespace.
+
+To designate that a ClusterDeployment is authorized to consume Agents from a
+particular InstallEnv, a reference to that ClusterDeployment must be present
+in an array on the InstallEnv. The reference proves that someone with write
+access to the InstallEnv approves of associating it with the cluster.
+
+The restriction will be enforced by the agent-related controllers at the time
+of matching Agents to a particular ClusterDeployment. The controllers will
+still have access to all namespaces.
```

Comment:
Is this late binding complication necessary? Do our initial customers have different people in the roles of Cluster Creator and Infra Owner? Do they have hardware that is not pre-allocated to pre-planned clusters for which namespaces could be created in advance?

Reply:
That is a main use case that we're working towards